pypose/module/particle_filters.py

```diff
@@ -0,0 +1,184 @@
+        particle_number (:obj:`Int`, optional): The number of particle. Default: ``1000``
+
+    A non-linear system can be described as
+
+    .. math::
+        \begin{aligned}
+            \mathbf{x}_{k+1} &= \mathbf{f}(\mathbf{x}_k, \mathbf{u}_k, t_k) + \mathbf{w}_k,
+            \quad \mathbf{w}_k \sim \mathcal{N}(\mathbf{0}, \mathbf{Q})  \\
+            \mathbf{y}_{k} &= \mathbf{g}(\mathbf{x}_k, \mathbf{u}_k, t_k) + \mathbf{v}_k,
+            \quad \mathbf{v}_k \sim \mathcal{N}(\mathbf{0}, \mathbf{R})
+        \end{aligned}
+
+    Warning:
+        Don't introduce noise in ``System`` methods ``state_transition`` and ``observation``
+        for filter testing, as those methods are used for automatically linearizing the system
+        by the parent class ``pypose.module.System``, unless your system model explicitly
+        introduces noise.
+
+    Note:
+        Implementation is based on Section 15.3 of this book
+
+        * Dan Simon, `Optimal State Estimation: Kalman, H∞, and Nonlinear Approaches
+          <https://onlinelibrary.wiley.com/doi/epdf/10.1002/0470045345.fmatter>`_,
+          Cleveland State University, 2006
+    '''
+
+    def __init__(self, model, Q=None, R=None, particle_number=100):
+        super().__init__()
+        self.set_uncertainty(Q=Q, R=R)
+        self.model = model
+        self.particle_number = particle_number
+
+    def generate_particle(self, x, Q):
+        r'''
+        randomly generate particle
+        '''
+
+        x_sample = torch.distributions.MultivariateNormal(x, Q).sample(
+            (self.particle_number,))
+
+        return x_sample
+
+    def relative_likelihood(self, y_estimate, y, R):
+        r'''
+        Compute the relative likelihood
+        '''
+
+        rv = stats.multivariate_normal(y, R)
+        q = rv.pdf(y_estimate) + 0.01
+        q = q / np.sum(q)
+
+        return torch.from_numpy(q)
+
+    def refine_particles(self, q, x):
+        r'''
+        Refine the set of a posteriori particles
+        '''
+
+        x_new = torch.zeros(self.particle_number, x.shape[1])
+        for n in range(self.particle_number):
+            r = torch.rand(1)[0]
+            sum_q = 0.
+
+            for j in range(self.particle_number):
+                sum_q += q[j]
+                if sum_q >= r:
+                    x_new[n] = x[j]
+                    break
+
+        return x_new
+
+    def compute_conv(self, x_estimate, x_particle):
+        r'''
+        compute covariance
+
+        Args:
+            x_estimate (:obj:`Tensor`): System estimate.
+            x_particle (:obj:`Tensor`):  Particle of estimate.
+
+        Return:
+            out (:obj:`Tensor`): the covariance of  prior state estimation .
+        '''
+
+        e = torch.sub(x_particle, x_estimate).unsqueeze(2)
+        p_estimate = torch.sum(torch.bmm(e, e.permute(0, 2, 1)),
+                               dim=0) / (self.particle_number - 1)
+
+        return p_estimate
```

Comment:
follow the implementation of UKF [pull request](https://github.com/HongLouyemeng/pypose/pull/5)